### PR TITLE
[cuDNN SDPA] fix bias sharding check and dbias calculation

### PR DESCRIPTION
**Pull request by Cjkkkk**
* only check bias batch/num_head sharding spec if present. Both dims could be broadcast.
* dbias calculation is incorrect in spmd and all_reduce is required.

### CODE REVIEW

**Comment by Cjkkkk** [comment]
@superbobry could you help review this? Thanks.

**Review by superbobry — APPROVED** [review]

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 1]
```diff
@@ -713,7 +713,8 @@ def _check_qkv_bias_mask_spec(
   # check bias spec
   if bias_spec:
     *bias_batch_spec, bias_num_head_spec, bias_q_seq_spec, bias_kv_seq_spec = bias_spec
-    if bias_batch_spec != batch_spec or bias_num_head_spec != num_head_spec:
+    if any(bias_batch_spec) and bias_batch_spec != batch_spec or \
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:716]: I modify the `test_sdpa_broadcast_bias_and_dbias` from single device to sharded test.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 2]
```diff
@@ -805,16 +806,26 @@ def _dot_product_attention_bwd_partition(
   out_shardings = _infer_bwd_output_sharding(mesh, arg_shapes, variadic_args)
   # args sharding
   arg_shardings = tuple([arg_i.sharding for arg_i in arg_shapes])
-  impl = functools.partial(
+  def sharded_impl(*args):
+    impl = functools.partial(
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: fixed.

**Comment by superbobry** [comment]
Please squash the commit so that we can merge.

**Comment by Cjkkkk** [comment]
Done, thanks!

**Pull request closed without merging**